android/src/main/java/cl/json/social/ShareIntent.java

```diff
@@ -167,6 +167,7 @@ public void open(ReadableMap options) throws ActivityNotFoundException {
             this.fileShare = getFileShare(options);
             if (this.fileShare.isFile()) {
                 Uri uriFile = this.fileShare.getURI();
+                this.getIntent().setAction(android.content.Intent.ACTION_SEND);
```

Comment:
What I don't understand is, this object is created from the start with ACTION_SEND, see https://github.com/react-native-share/react-native-share/pull/1191/files#diff-13df424eeb2513e4ed02873c2c63e6c21692ef91233afac79cb28dceb3a8ffa0R47

The only reason it could be overridden (by searching in file for ACTION_) is if you have key 'urls' (which is exclusive of 'url') - then it sends to ACTION_SEND_MULTIPLE or if you Share (not ShareSingle!) where it calls openChooserIntent and sets it to ACTION_VIEW

This is deduced by noting WhatsAppShare extends ShareSingle, which is where openChooserIntent is overridden from ShareIntent::openChooserIntent and *does not* override the action:

https://github.com/react-native-share/react-native-share/blob/d018c3715ad60798e3c389a51eb71bd25a18209c/android/src/main/java/cl/json/social/SingleShareIntent.java#L62-L66

So...I can't see how this should be necessary? Have you tried a `System.err.println` in the java code for ShareSingle.java and had it spit out the Intent immediately prior to this line https://github.com/react-native-share/react-native-share/blob/d018c3715ad60798e3c389a51eb71bd25a18209c/android/src/main/java/cl/json/social/SingleShareIntent.java#L91 to see what the Intent contents are when you call WhatsAppShare ?

I'd be surprised if it wasn't already ACTION_SEND even with your change commented out - it seems like it should be based on my read of the code

Reply:
![2022-04-25 122558](https://user-images.githubusercontent.com/4749790/165121572-9d978cde-9bf7-4ae3-b594-d21f9e6502af.png)

@mikehardy I did the 'System.err.println("ACTION: " + this.getIntent().getAction());' and you can see that it has ACTION.VIEW

Reply:
It is strange that a "NOT INSTALLED" appears up there and I do have whatsapp installed.

Reply:
@edugolive you are on to something! In the NOT_INSTALLED case it resets intent to ACTION_VIEW! We're closer to a root cause now. I did not consider that code branch before.

Why is it showing up as not installed? Solve that, and your current change goes away, but we have a real fix that is targeted. Does WhatsApp have a different app for Android 12 :thinking:  ?

Reply:
<queries>
    <intent>
        <action android:name="android.intent.action.SEND" />
        <data android:mimeType="image/jpeg" />
    </intent>
</queries>

Reply:
"
<queries>
        <intent>
            <action android:name="android.intent.action.SEND" />
            <data android:mimeType="image/jpeg" />
        </intent>
    </queries>"

Reply:
> <queries>
        <intent>
            <action android:name="android.intent.action.SEND" />
            <data android:mimeType="image/jpeg" />
        </intent>
    </queries>

Reply:
Exactly that! This is strictly required for Android 12 - https://react-native-share.github.io/react-native-share/docs/install#adding-queries-for-the-android-necessary-for-sdk--30

Before closing this - perhaps the whole idea of the PR could be changed. Right now it's "this is not working on Android 12, let's force it", but the root cause was really "I did not forward port to Android 12 correctly". And the root cause happens because we all have too much to do to remember everything, this will happen to others. Is probably happening right now to others!

Perhaps the real PR could be:
- If API is >= Android 12
- And If the build is a debug build (signaling it is for development)
- And if ShareSingle is called
- And if we hit that NOT INSTALLED conditional block
- We somehow report this back to the javascript layer and do a console.warn about the possible issue so developers get a yellow box

The whole idea is to pop this up to developers so it's obvious.

We handle the response callback from native here:

https://github.com/react-native-share/react-native-share/blob/d018c3715ad60798e3c389a51eb71bd25a18209c/src/index.tsx#L98

We send status of just "true" if the share was accepted by the system:
https://github.com/react-native-share/react-native-share/blob/d018c3715ad60798e3c389a51eb71bd25a18209c/android/src/main/java/cl/json/RNShareModule.java#L51

Perhaps we could overload that return value from simple true/false to also potentially be a string containing a message indicating that the package wasn't found, and if we detect that we can do the console.warn ?

Thoughts?

Reply:
But @mikehardy, what should I do?
I don't have enough knowledge to implement these changes.

Reply:
Okay - I think my implementation sketch is good but I certainly understand. I know how to do it but don't have the time right now so I peeled it to a separate issue based on the comment I made. I'll close this then as I think we got the root cause :+1:

Reply:
Perfect thank you.